Docs: add merge section to reference (#4308)
treeverse/lakeFS: docs/reference/merge.md

---
layout: default
title: Merging Reference
description: Using lakeFS, you can merge different commits and references into a branch. The purpose of this document is to explain how to use this feature.
parent: Reference
has_children: false
---

# Merge

> lakeFS allows merging a _merge source_ (a commit/reference) into a _merge destination_ (a **branch**). 

Reviewer: ```suggestion
The merge operation in lakeFS is similar to Git. It incorporates changes from a _merge source_ (a commit/reference) into a _merge destination_ (a **branch**). 
```

> ## How it works?

Reviewer 2: ```suggestion
## How does it work?
```

or
```suggestion
## What does a merge do?
```

> lakeFS first finds the [merge base](https://git-scm.com/docs/git-merge-base#_description) the nearest common parent of the two commits.

Reviewer 2: ```suggestion
lakeFS first finds the [merge base](https://git-scm.com/docs/git-merge-base#_description): the nearest common ancestor of the two commits.
```

|      A      |       A       |         A          |     A      | Unchanged file                                 |
|      A      |       B       |         B          |     B      | Files changed on both sides in same way        |
|      A      |       B       |         C          |  conflict  | Files changed on both sides differently        |
|      A      |       A       |         B          |     B      | File changed only on one branch                |
|      A      |       B       |         A          |     B      | File changed only on one branch                |
|      A      |       X       |         X          |     X      | Files deleted on both sides                    |
|      A      |       B       |         X          |  conflict  | File changed on one side, deleted on the other |
|      A      |       X       |         B          |  conflict  | File changed on one side, deleted on the other |
|      A      |       A       |         X          |     X      | File deleted on one side                       |
|      A      |       X       |         A          |     X      | File deleted on one side                       |

> ## Merging Strategies

Reviewer 2: ```suggestion
## Merge Strategies
```

The [API](./api.md) and [`lakectl`](./commands.md#lakectl-merge) allow passing an optional `strategy` flag with the following values:

### `source-wins`

In case of a conflict, merge will pick the source objects.

#### Example

```bash
lakectl merge lakefs://example-repo/validated-data lakefs://example-repo/production --strategy source-wins
```

> When a merge conflict arise, the conflicting objects in the `validated-data` branch will be chosen and found in `production`.

Reviewer 2: ```suggestion
When a merge conflict arises, the conflicting objects in the `validated-data` branch will be chosen to end up in `production`.
```

### `dest-wins`

In case of a conflict, merge will pick the destination objects.

#### Example

```bash
lakectl merge lakefs://example-repo/validated-data lakefs://example-repo/production --strategy dest-wins
```

> When a merge conflict arise, the conflicting objects in the `production` branch will be chosen and found in `validated-data`. The `production` branch will not be affected by object changes from `validated-data` conflicted objects.

Reviewer 2: ```suggestion
When a merge conflict arises, the conflicting objects in the `production` branch will be chosen to end up in `validated-data`. The `production` branch will not be affected by object changes from `validated-data` conflicting objects.
```

> The strategy will affect all conflicting objects in the merge if it is set. At the moment, it is not possible to treat conflicts individually.

Reviewer 2: ```suggestion
The strategy will affect all conflicting objects in the merge if it is set. Currently it is not possible to treat conflicts individually.
```

> As a format-agnostic system, lakeFS currently merges by complete files. Format-specific and

Reviewer 2: Do we have a link to this item on the roadmap, or an issue?

Author: It's actually not on the roadmap (at least I couldn't find it), but we have this [issue](https://github.com/treeverse/lakeFS/issues/3639) though

Reviewer 2: So please let's link to that!  (and write on that issue to update this doc when we fix the issue, otherwise that link will end up looking weird...)
